### PR TITLE
sql: Keep queries in SHOW QUERIES/SESSIONS until results are sent, add last_active_query column

### LINKED ISSUE #17541: sql: `SHOW QUERIES` output doesn't list queries with results being sent
**BUG REPORT**

1. Please supply the header (i.e. the first few lines) of your most recent
   log file **for each node in your cluster**. On most unix-based systems
   running with defaults, this boils down to the output of

     grep -F '[config]' cockroach-data/logs/cockroach.log

```
~/Downloads/cockroach-v1.1-alpha.20170803.darwin-10.9-amd64$ grep -F '[config]' cockroach-data/logs/cockroach.log
I170809 13:52:35.963735 1 util/log/clog.go:975  [config] file created at: 2017/08/09 13:52:35
I170809 13:52:35.963735 1 util/log/clog.go:975  [config] running on machine: JESSEs-MBP
I170809 13:52:35.963735 1 util/log/clog.go:975  [config] binary: CockroachDB CCL v1.1-alpha.20170803 (darwin amd64, built 2017/08/03 17:19:14, go1.8.3)
I170809 13:52:35.963735 1 util/log/clog.go:975  [config] arguments: [./cockroach start --insecure --host=localhost --port=26258 --http-port=8081 --join=localhost:26257]
```

2. Please describe the issue you observed:

- What did you do?

I ran ycsb against a 3-node local cluster. Then I opened the sql shell and ran `SHOW CLUSTER SETTINGS;`.

- What did you expect to see?

Because I ran ycsb with 10 concurrent clients, I expected to see 11 sessions (10 from ycsb and one from my sql shell) with the active queries listed for each connection.

- What did you see instead?

I saw 11 sessions, but the `active_queries` were usually empty, except for the sql client session:

```
root@:26258/> SHOW CLUSTER SESSIONS;
+---------+----------+-----------------+------------------+-------------------------+----------------------------------+----------------------------------+--------+
| node_id | username | client_address  | application_name |     active_queries      |          session_start           |        oldest_query_start        | kv_txn |
+---------+----------+-----------------+------------------+-------------------------+----------------------------------+----------------------------------+--------+
|       2 | root     | 127.0.0.1:54011 |                  | SHOW CLUSTER SESSIONS;  | 2017-08-09 12:05:37.614049+00:00 | 2017-08-09 12:06:51.963431+00:00 | NULL   |
|       1 | root     | 127.0.0.1:53998 |                  |                         | 2017-08-09 12:05:28.636798+00:00 | NULL                             | NULL   |
|       1 | root     | 127.0.0.1:53999 |                  |                         | 2017-08-09 12:05:28.638399+00:00 | NULL                             | NULL   |
|       1 | root     | 127.0.0.1:54006 |                  |                         | 2017-08-09 12:05:28.635602+00:00 | NULL                             | NULL   |
|       1 | root     | 127.0.0.1:54001 |                  |                         | 2017-08-09 12:05:28.635755+00:00 | NULL                             | NULL   |
|       1 | root     | 127.0.0.1:54005 |                  |                         | 2017-08-09 12:05:28.635947+00:00 | NULL                             | NULL   |
|       1 | root     | 127.0.0.1:54003 |                  |                         | 2017-08-09 12:05:28.6361+00:00   | NULL                             | NULL   |
|       1 | root     | 127.0.0.1:53993 |                  |                         | 2017-08-09 12:05:08.102355+00:00 | NULL                             | NULL   |
|       1 | root     | 127.0.0.1:54002 |                  |                         | 2017-08-09 12:05:28.635783+00:00 | NULL                             | NULL   |
|       1 | root     | 127.0.0.1:54004 |                  |                         | 2017-08-09 12:05:28.635899+00:00 | NULL                             | NULL   |
|       1 | root     | 127.0.0.1:54000 |                  |                         | 2017-08-09 12:05:28.637144+00:00 | NULL                             | NULL   |
+---------+----------+-----------------+------------------+-------------------------+----------------------------------+----------------------------------+--------+
(11 rows)
```

After repeatedly re-running `SHOW CLUSTER SESSIONS`, I occasionally saw a few `select` statements listed:

```
root@:26258/> SHOW CLUSTER SESSIONS;
+---------+----------+-----------------+------------------+---------------------------------------------------------------------+----------------------------------+----------------------------------+--------------------------------------+
| node_id | username | client_address  | application_name |                           active_queries                            |          session_start           |        oldest_query_start        |                kv_txn                |
+---------+----------+-----------------+------------------+---------------------------------------------------------------------+----------------------------------+----------------------------------+--------------------------------------+
|       2 | root     | 127.0.0.1:54011 |                  | SHOW CLUSTER SESSIONS;                                              | 2017-08-09 12:05:37.614049+00:00 | 2017-08-09 12:06:52.842735+00:00 | NULL                                 |
|       1 | root     | 127.0.0.1:53998 |                  |                                                                     | 2017-08-09 12:05:28.636798+00:00 | NULL                             | NULL                                 |
|       1 | root     | 127.0.0.1:53999 |                  | SELECT * FROM ycsb.usertable WHERE ycsb_key = 5220360707986222959;  | 2017-08-09 12:05:28.638399+00:00 | 2017-08-09 12:06:52.8428+00:00   | NULL                                 |
|       1 | root     | 127.0.0.1:54006 |                  | SELECT * FROM ycsb.usertable WHERE ycsb_key = 2539364586279394330;  | 2017-08-09 12:05:28.635602+00:00 | 2017-08-09 12:06:52.849393+00:00 | 9ad2697b-9ab6-4efd-8a3c-3f2464675b66 |
|       1 | root     | 127.0.0.1:54001 |                  |                                                                     | 2017-08-09 12:05:28.635755+00:00 | NULL                             | NULL                                 |
|       1 | root     | 127.0.0.1:54005 |                  |                                                                     | 2017-08-09 12:05:28.635947+00:00 | NULL                             | NULL                                 |
|       1 | root     | 127.0.0.1:54003 |                  | SELECT * FROM ycsb.usertable WHERE ycsb_key = 7166771442326333162;  | 2017-08-09 12:05:28.6361+00:00   | 2017-08-09 12:06:52.850751+00:00 | 34abf4f4-366c-4701-bc69-5a49a3396e70 |
|       1 | root     | 127.0.0.1:53993 |                  | SELECT * FROM ycsb.usertable WHERE ycsb_key = 7120461415806568161;  | 2017-08-09 12:05:08.102355+00:00 | 2017-08-09 12:06:52.850698+00:00 | e3d642b9-51a4-47dc-a2f3-5ae4fde8d386 |
|       1 | root     | 127.0.0.1:54002 |                  |                                                                     | 2017-08-09 12:05:28.635783+00:00 | NULL                             | NULL                                 |
|       1 | root     | 127.0.0.1:54004 |                  |                                                                     | 2017-08-09 12:05:28.635899+00:00 | NULL                             | NULL                                 |
|       1 | root     | 127.0.0.1:54000 |                  |                                                                     | 2017-08-09 12:05:28.637144+00:00 | NULL                             | NULL                                 |
+---------+----------+-----------------+------------------+---------------------------------------------------------------------+----------------------------------+----------------------------------+--------------------------------------+
(11 rows)
```

Also, `insert` queries were occasionally listed. `yscb` writes a lot less frequently than it reads, but I definitely expected more writes to show up.

### PR DESCRIPTION
**Pull request by itsbilal**
Asynchronous/parallel queries are still deregistered at end of
execution, and so are `SHOW TRANSACTION STATUS`es (a special case in the executor).

Fixes #17541 . 

Also add a new column to the output of `SHOW SESSIONS` : `last_active_query`, which is the last SQL statement that ran on that session.

### CODE REVIEW

File: pkg/sql/executor.go [hunk 1]
```diff
@@ -1126,10 +1126,12 @@ func (e *Executor) execStmtsInCurrentTxn(
 		queryMeta.ctx = txnState.Ctx
 		queryMeta.ctxCancel = txnState.cancel
 
+		// For parallel/async queries, we deregister queryMeta from these registries
+		// after execution finishes in the parallelizeQueue. SHOW TRANSACTION STATUSes
+		// are deregistered in their own code path. For all other queries,
+		// we deregister these in session.FinishPlan.
```
Comment (pkit) [inline pkg/sql/executor.go]: Hi,
what is the point in leaving two places of query cancellation when it's pretty clear right now that it should happen after sending results?

File: pkg/sql/executor.go [hunk 2]
```diff
@@ -1148,6 +1150,9 @@ func (e *Executor) execStmtsInCurrentTxn(
 		// always guaranteed to execute regardless of the current transaction state.
 		if _, ok := stmt.AST.(*parser.ShowTransactionStatus); ok {
 			res, err = runShowTransactionState(session)
+
+			session.removeActiveQuery(queryID)
+			e.cfg.QueryRegistry.deregister(queryID)
```
Comment (pkit) [inline pkg/sql/executor.go]: What is the point of dragging these both statements everywhere? At least combine them into one function? Or just cancel everything inside `FinishPlan`

**Comment by knz** [comment]
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/executor.go, line 1132 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17579#-KrCGzTYsxIz93nJBHHl-r1-1132:-KrFfykGJY62n2tvnvOR:b-dm6oam) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d2213b9b73e8486c87aefd00faa4f9aa18c2bec/pkg/sql/executor.go#L1132)):*
<details><summary><i>Previously, pkit wrote…</i></summary><blockquote>

Hi,
what is the point in leaving two places of query cancellation when it's pretty clear right now that it should happen after sending results?

</blockquote></details>

Because parallel-run queries run *after* FinishPlan() has executed. They don't have results.

@itsbilal: please extend the comment to explain the difference between the two. I also made a double take.

---

*[pkg/sql/executor.go, line 1155 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17579#-KrCGzTYsxIz93nJBHHl-r1-1155:-KrFgmqQ94lihKCfKVMp:bqg8w43) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d2213b9b73e8486c87aefd00faa4f9aa18c2bec/pkg/sql/executor.go#L1155)):*
<details><summary><i>Previously, pkit wrote…</i></summary><blockquote>

What is the point of dragging these both statements everywhere? At least combine them into one function? Or just cancel everything inside `FinishPlan`
</blockquote></details>

Agreed, I think `FinishPlan` also runs for SHOW TRANSACTION STATUS.

---

*[pkg/sql/session.go, line 559 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17579#-KrFgxnH7p4IgrE-lFy0:-KrFgxnH7p4IgrE-lFy1:b-i0aorc) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d2213b9b73e8486c87aefd00faa4f9aa18c2bec/pkg/sql/session.go#L559)):*
> ```Go
> 	// active queries from this session.
> 	s.mu.Lock()
> 	for queryID, queryMeta := range s.mu.ActiveQueries {
> ```

Okay here I'm very confused:
1) I don't think there be more than one running non-parallel query when `FinishPlan` is executed. Why a loop? At least there's a `break` missing.
2) this lookup seems inefficient. There must be a better/faster way.

I think `ActiveQueries` by construction necessarily only contains `isParallel` queries for every item, or every item but the last or the first, and the only non-isParallel item should be found either in first position or at last. Can you check this? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17579)*

**Comment by pkit** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/executor.go, line 1132 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17579#-KrCGzTYsxIz93nJBHHl-r1-1132:-KrGQvhTPuPwtKrU3RBf:btjiwz) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d2213b9b73e8486c87aefd00faa4f9aa18c2bec/pkg/sql/executor.go#L1132)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Because parallel-run queries run *after* FinishPlan() has executed. They don't have results.

@itsbilal: please extend the comment to explain the difference between the two. I also made a double take.
</blockquote></details>

Correct me if I'm wrong, but everything runs from `execStmtsInCurrentTxn` anyway (all other functions like `execStmtsInOpenTxn`, etc.), and `FinishPlan` is called anyway after that. Therefore it looks like all the cancellation can be consolidated inside `FinishPlan`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17579)*

**Comment by knz** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/executor.go, line 1132 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17579#-KrCGzTYsxIz93nJBHHl-r1-1132:-KrGTCkOhSswf-GPxl-A:b-313g37) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d2213b9b73e8486c87aefd00faa4f9aa18c2bec/pkg/sql/executor.go#L1132)):*
<details><summary><i>Previously, pkit (Constantine Peresypkin) wrote…</i></summary><blockquote>

Correct me if I'm wrong, but everything runs from `execStmtsInCurrentTxn` anyway (all other functions like `execStmtsInOpenTxn`, etc.), and `FinishPlan` is called anyway after that. Therefore it looks like all the cancellation can be consolidated inside `FinishPlan`.
</blockquote></details>

Parallel-run queries are spawn in a separate goroutine. execStmtsInCurrentTxn finishes before the parallel queries complete.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17579)*

**Comment by knz** [comment]
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/session.go, line 559 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17579#-KrFgxnH7p4IgrE-lFy0:-KrHDjfvMXKdSvPlgYRZ:b-c2sau) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d2213b9b73e8486c87aefd00faa4f9aa18c2bec/pkg/sql/session.go#L559)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

1.  `execStmtsInCurrentTxn` could be called with multiple statements, and it returns a slice of multiple `Result`s. `FinishPlan` is only called once for a request (and it's called from pgwire after all results are sent). If I just deregister the last statement run, any previous statements in that same request will stay in the registries permanently.
2. Unless we store non-parallel queries separately, I'm not sure how we could get away without a loop. Also this is a map, not a slice (so we can't just look at the last few elements. Otherwise, yes, the non-parallel statements will comprise the latest additions to the map).
</blockquote></details>

I'm concerned by the fact that FinishPlan is in the hot path (every query). We shouldn't have all-map iterations in there.
Please investigate adding the appropriate data structure that would make this short even if there are parallel queries in the background.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17579)*

**Comment by itsbilal** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/session.go, line 559 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17579#-KrFgxnH7p4IgrE-lFy0:-KrWNBEVMbx0jNwrmhmH:bx9w1oy) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d2213b9b73e8486c87aefd00faa4f9aa18c2bec/pkg/sql/session.go#L559)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I'm concerned by the fact that FinishPlan is in the hot path (every query). We shouldn't have all-map iterations in there.
Please investigate adding the appropriate data structure that would make this short even if there are parallel queries in the background.
</blockquote></details>

The best I can think of (in this context, where we have to assume a query is synchronous until it's explicitly marked as parallel), is a shadow map that only has synchronous queries. By default, you'd put a query in both maps. Whenever a query is marked as async, you'd just delete it from the synchronous map. In `FinishPlan`, you'll still need an all-map iteration on that map so that you can delete() the matching entry in the bigger map.

This would reduce the size of this iteration down to the number of synchronous statements in the request you're `FinishPlan()`ing. If that sounds like a good idea, I can go ahead and implement that. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17579)*

**Comment by itsbilal** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/session.go, line 559 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17579#-KrFgxnH7p4IgrE-lFy0:-KrWQn0p__CvcDE1lOpY:b-20xiqi) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d2213b9b73e8486c87aefd00faa4f9aa18c2bec/pkg/sql/session.go#L559)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

The best I can think of (in this context, where we have to assume a query is synchronous until it's explicitly marked as parallel), is a shadow map that only has synchronous queries. By default, you'd put a query in both maps. Whenever a query is marked as async, you'd just delete it from the synchronous map. In `FinishPlan`, you'll still need an all-map iteration on that map so that you can delete() the matching entry in the bigger map.

This would reduce the size of this iteration down to the number of synchronous statements in the request you're `FinishPlan()`ing. If that sounds like a good idea, I can go ahead and implement that. 
</blockquote></details>

Actually - we can take advantage of the fact that `setQueryExecutionMode` is called from the main session goroutine, even for async queries. So the moment we find out isParallel = true, we can just chop off the last added element from a slice of synchronous queries. Moving from a map to a slice will also be more space-efficient (and more performant in general for `range` iterations?). I've made this change - PTAL!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17579)*

**Comment by knz** [comment]
You need to rebase and also update the logic tests `crdb_internal` and `explain`.

While you're doing this please also consider the following: I am not sure we need `execCfg.QueryRegistry`.
My reasoning is as follows:

- maintaining a Go map on the "hot path" of the executor adds overhead
- overhead on the "hot path" is only acceptable either **if functionally necessary** or **if it enables a complexity class difference elsewhere on a hot path** (e.g. it's ok to access an extra data structure on a hot path if there's no other place to maintain it, or if it's used somewhere else on a hot path for faster something)

- the registry is not functionally necessary:
   - we do not need the registry to guarantee query ID uniqueness, the ID is unique by construction
   - we can find a query by ID by scanning all sessions and then every query in each session
- the registry doesn't participate in any hot path:
  - the registry is not *used* in the executor
  - we do not need a query lookup by ID to be fast (it's a DBA command likely to be run interactively), so the query registry doesn't participate in the hot path somewhere else
- therefore, the registry is not necessary

did I miss something?

---

Reviewed 1 of 2 files at r3, 7 of 7 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17579#-:-KrefYiLT0fhcXTl7yJE:b-l9pacl)*

**Comment by knz** [comment]
(The rest of your changes LGTM.)

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17579#-:-KrehMP2FRL9x-TQYOhb:b5fgjyi)*

**Comment by itsbilal** [comment]
@knz The query registry is used to look up queries for `CANCEL QUERY`, since that command only gets a query ID. So I'd argue that it's still a reasonable registry to have. 

Thanks for the review! Updating tests now.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17579#-:-KrfI_qCRVzg8uMN-_yt:b-o30e4w)*

**Comment by knz** [comment]
I repeat, as far as I can see the query ID is already unique across sessions. CANCEL QUERY can go through all session and search for the ID in each of them.

**Comment by itsbilal** [comment]
Removed the query registry in the latest commit. PTAL!

**Comment by knz** [comment]
:lgtm_strong:

---

Reviewed 7 of 7 files at r5, 8 of 9 files at r6, 4 of 4 files at r7.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17579#-:-Krfi59AvhiClGKkzu0W:b-jbsf8z)*

**Pull request merged**